### PR TITLE
systemd: Avoid path_namespace watches with org.freedesktop.problems

### PR DESCRIPTION
**Pull request by mvollmer**
o.fd.problems does not have a ObjectManager and we might not want to create a DBusProxies object for it, which relies on emulating a ObjectManager via Introspection.

Thus, let's create the array of DBusProxy:s explicitly.

### CODE REVIEW

**Comment by mvollmer** [comment]
This is still inefficient, since it refetches all information on each call to loadEntry.  However, this call only happens when people navigate to the details of a single message, and thus doesn't impact the main journal page.  We might be okay with living with it.

File: pkg/systemd/logDetails.jsx [hunk 1]
```diff
@@ -93,29 +110,26 @@ export class LogEntry extends React.Component {
...
+        const client = cockpit.dbus('org.freedesktop.problems', { superuser: "try" });
+        const service = client.proxy('org.freedesktop.Problems2', '/org/freedesktop/Problems2');
+
+        get_problems(service)
+                .then(problems => {
+                    const fields = [entry.PROBLEM_DIR, entry.PROBLEM_DUPHASH, entry.PROBLEM_UUID];
+                    let path = null;
+                    Object.keys(problems).some(pth => {
+                        const p = problems[pth];
+                        if (p && (fields.indexOf(p.ID) > 0 || fields.indexOf(p.UUID) || fields.indexOf(p.Duphash))) {
```
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:122]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221216-143043-67e35acc-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx [hunk 2]
```diff
@@ -93,29 +110,26 @@
...
+                            path = p;
+                            return true;
+                        } else {
+                            return false;
+                        }
```
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:127]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221216-143043-67e35acc-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by martinpitt — APPROVED** [review]
When looking at this code, I think "it's high time to start using `await` :grin: 

Honestly, I'm ok with that. The structure isn't nice, but at least it eliminates the implicit assumption about the ObjectManager emulation, which I think is a good thing. I would prefer doing this without the expensive proxies at all, and just use a plain dbus.client, but this is already a good enough improvement. Thanks!

PR is still draft, but feel free to land as-is. This will "collide" with #18039 in the sense of that it will probably make the ABRT tests pass, and thus "break" the exfail. Let's coordinate on Monday!

**Comment by mvollmer** [comment]
I want to have one more look at how often this code actually runs, and whether we can easily avoid leaking all these proxies on each render.

**Comment by martinpitt** [comment]
Wow, [what is that crap](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-073136-9aa9eeeb-centos-8-stream/log.html)?? Let's do a round of retries.

File: pkg/systemd/logDetails.jsx [hunk 3]
```diff
@@ -90,32 +107,32 @@ export class LogEntry extends React.Component {
...
-                                return false;
-                            }
-                        });
-
-                        this.setState({ entry: entry, loading: false, error: "", problemPath: path, abrtService: service });
-                    })
-                    .catch(err => this.setState({ entry: entry, loading: false, error: err.toString() }));
-        });
+        if (this.problems_client)
+            this.problems_client.close();
```
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:115]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-103616-9aa9eeeb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx [hunk 4]
```diff
@@ -90,32 +107,32 @@
...
+        this.problems_client = cockpit.dbus('org.freedesktop.problems', { superuser: "try" });
+
+        const service = this.problems_client.proxy('org.freedesktop.Problems2', '/org/freedesktop/Problems2');
+        get_problems(service)
+                .then(problems => {
+                    const fields = [entry.PROBLEM_DIR, entry.PROBLEM_DUPHASH, entry.PROBLEM_UUID];
+                    let path = null;
+                    Object.keys(problems).some(pth => {
+                        const p = problems[pth];
+                        if (p && (fields.indexOf(p.ID) > 0 || fields.indexOf(p.UUID) || fields.indexOf(p.Duphash))) {
```
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:125]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-103616-9aa9eeeb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx [hunk 5]
```diff
@@ -90,32 +107,32 @@
...
+                            path = p;
+                            return true;
+                        } else {
+                            return false;
+                        }
```
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:130]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-103616-9aa9eeeb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx [hunk 6]
```diff
@@ -90,32 +107,32 @@
...
+                    });
+
+                    this.setState({ entry: entry, loading: false, error: "", problemPath: path, abrtService: service });
+                })
+                .catch(err => this.setState({ entry: entry, loading: false, error: err.toString() }));
```
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:135]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-103616-9aa9eeeb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
pybridge scenario fixes landed! So I'm going to trigger the run here -- I expect the ABRT tests to fail now, as they should unexpectedly succeed.

**Comment by martinpitt** [comment]
Aha! This makes [three abrt tests](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-220331-9aa9eeeb-fedora-36-pybridge/log.html) pass (and hence shown as failure, as they are decorated with `@todoPybrige`). The two main ones still fail, though.

I'm not sure why `TestAccounts.testBasic` suddenly starts succeeding. It [previously](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18039-20221219-175300-efaf2b28-fedora-36-pybridge/log.html) timed out on `passwd -S root`. However, it's [wonky](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18039-20221216-142401-b4036bbe-fedora-36-pybridge/log.html) in earlier runs, so let's take out the `expectedFailure` and fix it as a "normal" flake.

And finally, [TestAutoUpdates.testBasic](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-220331-9aa9eeeb-fedora-36-pybridge/log.html#150) triggered a [bridge crash](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221219-220331-9aa9eeeb-fedora-36-pybridge/TestAutoUpdates-testBasic-fedora-36-127.0.0.2-2401-FAIL.log.gz), this needs more message filtering.

**Comment by martinpitt** [comment]
@mvollmer : I updated the decorators for the three fixed tests, and added a message filter for the above crash. I left TestAccounts alone, if it fails again, I'll mark it as flaky.

File: pkg/systemd/logDetails.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/systemd/logDetails.jsx:115]: Yes, exactly.  Closing the client wil close the dbus-json3 channel, which is the easiest way to release all resources.

File: pkg/systemd/logDetails.jsx [hunk 7]
```diff
@@ -77,6 +77,23 @@ const LogDetails = ({ entry }) => {
     );
 };
 
+function get_problems(service) {
+    return service.wait()
+            .then(() => {
+                return service.GetProblems(0, {})
+                        .then(paths => {
+                            const proxies = paths.map(p => service.client.proxy("org.freedesktop.Problems2.Entry", p));
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/systemd/logDetails.jsx:85]: The rest of the code expect real DBusProxy objects, and constructing one isn't all that bad.  It results in a `watch` request to the bridge, which is more or less just a `GetAll` call, plus one time Introspection.

File: pkg/systemd/logDetails.jsx [hunk 8]
```diff
@@ -90,32 +107,32 @@ export class LogEntry extends React.Component {
...
-
-                        this.setState({ entry: entry, loading: false, error: "", problemPath: path, abrtService: service });
-                    })
-                    .catch(err => this.setState({ entry: entry, loading: false, error: err.toString() }));
-        });
+        if (this.problems_client)
+            this.problems_client.close();
+        this.problems_client = cockpit.dbus('org.freedesktop.problems', { superuser: "try" });
+
+        const service = this.problems_client.proxy('org.freedesktop.Problems2', '/org/freedesktop/Problems2');
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/systemd/logDetails.jsx:118]: The rest of the code expects a real proxy.

File: pkg/systemd/logDetails.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:115]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-065325-e1c6e2ec-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:125]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-065325-e1c6e2ec-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:130]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-065325-e1c6e2ec-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:135]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-065325-e1c6e2ec-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:115]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-092207-882459d6-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:125]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-092207-882459d6-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:130]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-092207-882459d6-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:135]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-092207-882459d6-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
Meh, [TestAccounts](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18052-20221220-112740-882459d6-rhel-8-7/log.html#329) is *really* flaky, this is the 5th(!) retry.

**Review by martinpitt — APPROVED** [review]
Looks good, if we can *ever* get this green. Thanks!

**Comment by martinpitt** [comment]
I'm going to ignore the rhel-8-7 accounts failure, it's just too flaky and unrelated.

**Pull request merged**